go/payload-commit/service/service.go

```diff
@@ -0,0 +1,578 @@
+				"txPayload": txPayload,
+			}, slackutils.SeverityCritical, ServiceName)
+
+			return err
+		}
+
+		return nil
+	}
+
+	return nil
+}
+
+// HandleFinalizedPayloadCommitTask handles finalized payload commit task.
+func (s *PayloadCommitService) HandleFinalizedPayloadCommitTask(msg *datamodel.PayloadCommitFinalizedMessage) error {
+	log.Debug("handling finalized payload commit task")
+
+	// check if payload is already in cache
+	payloadCid, err := s.redisCache.GetSnapshotCidAtEpochID(context.Background(), msg.Message.ProjectID, msg.Message.EpochID)
+	if err != nil {
+		log.WithError(err).Error("failed to get payload cid from redis")
+
+		return err
+	}
+
+	var report *datamodel.SnapshotterStatusReport
+
+	// if stored snapshot cid does not match with finalized snapshot cid, fetch commit message from ipfs and store in local disk.
+	if payloadCid != msg.Message.SnapshotCID {
+		log.Debug("payload cid does not match with finalized snapshot cid, fetching snapshot commit message from ipfs")
+
+		dirPath := filepath.Join(s.settingsObj.LocalCachePath, msg.Message.ProjectID, "snapshots")
+		filePath := filepath.Join(dirPath, msg.Message.SnapshotCID+".json")
+
+		// create file, if it does not exist
+		err = os.MkdirAll(dirPath, os.ModePerm)
+		if err != nil {
+			log.WithError(err).Error("failed to create file")
+		}
+
+		// get payload commit message from ipfs and store it in output path
+		err = s.ipfsClient.GetPayloadCommitMessageFromIPFS(msg.Message.SnapshotCID, filePath)
+		if err != nil {
+			log.WithError(err).Error("failed to get payload commit message from ipfs")
+
+			go s.slackNotifier.NotifySlackWorkflow(map[string]interface{}{
+				"msg":   "failed to get snapshot from ipfs",
+				"error": err.Error(),
+			}, slackutils.SeverityError, ServiceName)
+
+			return err
+		}
+
+		report = &datamodel.SnapshotterStatusReport{
+			SubmittedSnapshotCid: payloadCid,
+			FinalizedSnapshotCid: msg.Message.SnapshotCID,
+			Missed:               true,
+		}
+
+		// unpin unfinalized snapshot cid from ipfs
+		err = s.ipfsClient.Unpin(payloadCid)
+		if err != nil {
+			log.WithError(err).WithField("cid", payloadCid).Error("failed to unpin snapshot cid from ipfs")
+		}
+	} else {
+		report = &datamodel.SnapshotterStatusReport{
+			SubmittedSnapshotCid: payloadCid,
+			FinalizedSnapshotCid: msg.Message.SnapshotCID,
+			Missed:               false,
+		}
+
+		outputPath := filepath.Join(s.settingsObj.LocalCachePath, msg.Message.ProjectID, "snapshots", msg.Message.SnapshotCID+".json")
+
+		data, err := json.Marshal(msg.Message)
+		if err != nil {
+			log.WithError(err).Error("failed to marshal payload commit message")
+		}
+
+		err = s.diskCache.Write(outputPath, data)
+		if err != nil {
+			log.WithError(err).Error("failed to write payload commit message to disk cache")
+		}
+	}
+
+	// generate report and store in redis
+	err = s.redisCache.AddSnapshotterStatusReport(context.Background(), msg.Message.EpochID, msg.Message.ProjectID, report)
```

Comment:
Right now, all data (whether it was correct or incorrect) is being stored in Redis reports. Does it make sense to store the correct data? 
Even for incorrect/missed snapshots which parts of report are relevant? Let's think a bit more and structure the report data efficiently.

Reply:
Not storing report when the snapshot cid generated by snapshotter matches with the finalized snapshot cid